CMake/AbseilHelpers.cmake

```diff
@@ -62,7 +62,107 @@ function(absl_library)
   endif()
 endfunction()
 
+#
+# CMake function to imitate Bazel's cc_library rule.
+#
+# Parameters:
+# NAME: name of target (see Note)
+# HDRS: List of public header files for the library
+# SRCS: List of source files for the library
+# DEPS: List of other libraries to be linked in to the binary targets
+# COPTS: List of private compile options
+# DEFINES: List of public defines
+# LINKOPTS: List of link options
+# PUBLIC: Add this so that this library will be exported under absl:: (see Note).
+# TESTONLY: When added, this target will only be built if user passes -DBUILD_TESTING=ON to CMake.
+#
+# Note:
+#
+# By default, absl_cc_library will always create a library named absl_internal_${NAME},
+# which means other targets can only depend this library as absl_internal_${NAME}, not ${NAME}.
+# This is to reduce namespace pollution.
+#
+# absl_cc_library(
+#   NAME
+#     awesome_lib
+#   HDRS
+#     "a.h"
+#   SRCS
+#     "a.cc"
+# )
+# absl_cc_library(
+#   NAME
+#     fantastic_lib
+#   SRCS
+#     "b.cc"
+#   DEPS
+#     absl_internal_awesome_lib # not "awesome_lib"!
+# )
+#
+# If PUBLIC is set, absl_cc_library will also create an alias absl::${NAME}
+# for public use in addition to absl_internal_${NAME}.
+#
+# absl_cc_library(
+#   NAME
+#     main_lib
+#   ...
+#   PUBLIC
+# )
+#
+# User can then use the library as absl::main_lib (although absl_internal_main_lib is defined too).
+#
+# TODO: Implement "ALWAYSLINK"
 
+function(absl_cc_library)
+  cmake_parse_arguments(ABSL_CC_LIB
+    "DISABLE_INSTALL;PUBLIC;TESTONLY"
+    "NAME"
+    "HDRS;SRCS;COPTS;DEFINES;LINKOPTS;DEPS"
+    ${ARGN}
+  )
+
+  if (NOT ABSL_CC_LIB_TESTONLY OR ABSL_RUN_TESTS)
+    set(_NAME "absl_internal_${ABSL_CC_LIB_NAME}")
+
+    # Check if this is a header-only library
+    if (ABSL_CC_LIB_SRCS)
+      set(_SRCS ${ABSL_CC_LIB_SRCS})
+      list(FILTER _SRCS INCLUDE REGEX "\.cc$")
+      list(LENGTH _SRCS ABSL_CC_LIB_SRCS_LEN)
+    else()
+      set(ABSL_CC_LIB_SRCS_LEN 0)
+    endif()
```

Comment:
Is it possible to have a header only lib with SRCS parameter not empty ?

Reply:
Why is it necessary to have a header-only lib with SRCS parameter? Does Bazel allow other targets to depend on `.h` files declared in `srcs` of a target? At least from what I can see, Abseil's `BUILD` files don't do that.

Reply:
Exactly, `INTERFACE` with `_SRCS` will be an error (i.e. an ill-formed header only lib).
So you can simply check if SRCS is empty to know if it's a `INTERFACE` or not:
```cmake
if ("${ABSL_CC_LIB_SRCS}" STREQUAL "")
  set(ABSL_CC_LIB_IS_INTERFACE 1)
else()
  set(ABSL_CC_LIB_IS_INTERFACE 0)
endif()
```
then later
```cmake
if(NOT ABSL_CC_LIB_IS_INTERFACE)
  add_library(${NAME} STATIC "")
  target_sources(${NAME} PRIVATE ${_SRCS})
  ...
else(NOT ABSL_CC_LIB_IS_INTERFACE)
  add_library(${NAME} INTERFACE)
endif()
```

Reply:
Just checking, so we don't need to check `ABSL_CC_LIB_SRCS_LEN` anymore right?

It seems like `target_sources` is also a command that is available only in fairly new version of CMake. (I myself still uses 3.11)

Reply:
So the correct usage of `target_sources` should be this?

```
      target_sources(${_NAME}
          PRIVATE ${ABSL_CC_LIB_SRCS}
          PUBLIC ${ABSL_CC_LIB_HDRS}
      )
```

Reply:
You should keep `${ABSL_CC_LIB_HDRS}` private. PUBLIC want absolute out of the source file (I'm currently trying to play with it) :confused:.

Reply:
Hi, I dig a little on this topic (cf comment in google/cctz#83 --> https://github.com/google/cctz/pull/83/files/6cd878965d04595624f512546f59682b165ccd00#diff-af3b638bc2a3e6c650974192a53c7291)
TLDR the correct usage seems (when there is no install rule defined like it is the case for this PR):
```cmake
      target_sources(${_NAME}
          PRIVATE ${ABSL_CC_LIB_SRCS}
          PUBLIC ${ABSL_CC_LIB_HDRS}
      )
```
**BUT** HDRS must use ${CMAKE_CURRENT_SOURCE_DIR}/...

Once you have install rules you need to use generator also (note: no part of this PR however, (i'll do it in PR182 once I rebase it on this one...))

Reply:
Yeah I know i need to trash everything and rebuild on top of your PR.
no worry :+1: